core/src/main/java/org/apache/camel/kafkaconnector/CamelSourceTask.java

```diff
@@ -119,53 +119,63 @@ public void start(Map<String, String> props) {
         }
     }
 
+    private long remaining(long startPollEpochMilli, long maxPollDuration)  {
+        return maxPollDuration - (Instant.now().toEpochMilli() - startPollEpochMilli);
+    }
+
+
     @Override
     public synchronized List<SourceRecord> poll() {
-        long startPollEpochMilli = Instant.now().toEpochMilli();
+        final long startPollEpochMilli = Instant.now().toEpochMilli();
+
+        long remaining = remaining(startPollEpochMilli, maxPollDuration);
         long collectedRecords = 0L;
 
-        List<SourceRecord> records = new ArrayList<>();
-        while (collectedRecords < maxBatchPollSize && (Instant.now().toEpochMilli() - startPollEpochMilli) < maxPollDuration) {
-            Exchange exchange = consumer.receiveNoWait();
+        List<SourceRecord> records = null;
+        while (collectedRecords < maxBatchPollSize && remaining > 0) {
+            Exchange exchange = consumer.receive(remaining);
+            if (exchange == null) {
+                remaining = remaining(startPollEpochMilli, maxPollDuration);
+                continue;
+            }
+
+            if (records == null) {
```

Comment:
I would rather initialize `records` with and empty list check at the end if it is empty to return `null` for readability, than  getting (little?) gain in performance...

Reply:
+1